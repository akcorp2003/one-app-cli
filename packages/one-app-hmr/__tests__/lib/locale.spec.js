/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */
import chokidar from 'chokidar';
import createDebug from 'debug';

import {
  watchLanguagePackFileEvents,
  getModuleInfoFromLocalePath,
  createHotLanguagePacks,
  loadLanguagePacks,
} from '../../lib/locale';

jest.mock('path');
jest.mock('fs');
jest.mock('child_process', () => ({
  execSync: () => 'en-US',
}));

jest.mock('chokidar', () => {
  const mockChokidar = {
    watch: jest.fn(() => mockChokidar),
    on: jest.fn(() => mockChokidar),
  };

  return mockChokidar;
});
jest.mock('../../lib/webpack', () => ({
  loadWebpackMiddleware: async () => ({
    publish: jest.fn(),
  }),
}));

jest.mock('../../lib/utils', () => ({
  addLanguagePacksForModule: jest.fn(() => ['en-US']),
  removeModuleLanguagePack: jest.fn(),
  addModuleLanguagePack: jest.fn(),
}));
describe('locale build', () => {
  let originalDebugScope;

  beforeEach(() => {
    originalDebugScope = createDebug.disable();
  });

  afterEach(() => {
    console.log.mockClear();
    console.warn.mockClear();
    console.time.mockClear();
    chokidar.on.mock.calls = [];
    createDebug.enable(originalDebugScope);
  });

  const modules = ['sample-module-1', 'sample-module-2'];
  const consoleWarnSpy = jest.spyOn(console, 'warn')
    .mockImplementation();
  const consoleLogSpy = jest.spyOn(console, 'log')
    .mockImplementation();
  const consoleTimeSpy = jest.spyOn(console, 'time')
    .mockImplementation();
  const publish = jest.fn();
  test('gets module-name and locale from file path', () => {
    expect(getModuleInfoFromLocalePath('sample-module/locale/en-US.json')).toEqual(['sample-module', 'sample-module', 'en-US.json', 'en-us']);
  });
  describe('create hot reloaded language-packs', () => {
    test('on error log the error', () => {
      createHotLanguagePacks(modules, publish);
      chokidar.on.mock.calls[0][1]();
      expect(consoleWarnSpy.mock.calls)
        .toMatchSnapshot();
      expect(chokidar.on.mock.calls[0][0])
        .toBe('error');
    });
    test('on ready watches language packs', () => {
      createHotLanguagePacks(modules, publish);
      chokidar.on.mock.calls[1][1]();
      expect(consoleLogSpy.mock.calls)
        .toMatchSnapshot();
      expect(chokidar.on.mock.calls[1][0])
        .toBe('ready');
    });
  });
  describe('watchLanguagePackFileEvents', () => {
    test('on add hot reload language pack ', async () => {
      watchLanguagePackFileEvents(chokidar, publish);
      await chokidar.on.mock.calls[0][1]('sample-module/locale/en-US.json');
      expect(chokidar.on.mock.calls[0][0])
        .toBe('add');
      expect(consoleLogSpy.mock.calls)
        .toMatchSnapshot();
    });
    test('on change hot reload language pack ', async () => {
      watchLanguagePackFileEvents(chokidar, publish);
      await chokidar.on.mock.calls[1][1]('sample-module/locale/en-US.json');
      expect(chokidar.on.mock.calls[1][0])
        .toBe('change');
      expect(consoleLogSpy.mock.calls)
        .toMatchSnapshot();
    });
    test('on removed hot reload language pack', async () => {
      watchLanguagePackFileEvents(chokidar, publish);
      await chokidar.on.mock.calls[2][1]('sample-module/locale/en-US.json');
      expect(chokidar.on.mock.calls[2][0])
        .toBe('unlink');
      expect(consoleLogSpy.mock.calls)
        .toMatchSnapshot();
    });
  });
  test('language packs are loaded', async () => {
    const app = jest.fn();
    const languagePacks = ['en-US', 'en-GB'];
    const useLanguagePacks = true;
    await loadLanguagePacks(app, {
      languagePacks,
      useLanguagePacks,
      publish,
    });
    expect(chokidar.on.mock.calls[1][0])
      .toBe('ready');
    expect(consoleTimeSpy.mock.calls)
      .toMatchSnapshot();
  });
  test('language packs are not loaded', async () => {
    const app = jest.fn();
    const parameters = {
      languagePacks: [],
      useLanguagePacks: true,
      publish,
    };

    await loadLanguagePacks(app, parameters);
    expect(consoleTimeSpy.mock.calls.length)
      .toEqual(0);
  });
});
