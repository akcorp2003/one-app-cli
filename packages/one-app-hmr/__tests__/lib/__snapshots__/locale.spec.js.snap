// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`locale build create hot reloaded language-packs on error log the error 1`] = `
Array [
  Array [
    "[38;2;255;165;0m[1m  one-app-hmr ::[22m[39m %s",
    "[38;2;255;165;0m[38;2;152;251;152mlocale[39m[38;2;255;165;0m - Language pack watcher error: undefined[39m",
  ],
]
`;

exports[`locale build create hot reloaded language-packs on ready watches language packs 1`] = `
Array [
  Array [
    "[38;2;0;255;255m[1m  one-app-hmr ::[22m[39m %s",
    "[38;2;152;251;152mlocale[39m - Loaded language packs for \\"[38;2;255;165;0msample-module-1[39m\\": [32men-US[39m",
  ],
  Array [
    "[38;2;0;255;255m[1m  one-app-hmr ::[22m[39m %s",
    "[38;2;152;251;152mlocale[39m - Loaded language packs for \\"[38;2;255;165;0msample-module-2[39m\\": [32men-US[39m",
  ],
  Array [
    "[38;2;0;255;255m[1m  one-app-hmr ::[22m[39m %s",
    "[38;2;152;251;152mlocale[39m - Watching language packs for [38;2;255;165;0m\\"sample-module-1\\"[39m, [38;2;255;165;0m\\"sample-module-2\\"[39m",
  ],
]
`;

exports[`locale build language packs are loaded 1`] = `
Array [
  Array [
    "[38;2;255;228;196m[1m  one-app-hmr ::[22m[39m [38;2;152;251;152mlocale[39m - initializing",
  ],
]
`;

exports[`locale build watchLanguagePackFileEvents on add hot reload language pack  1`] = `
Array [
  Array [
    "[38;2;0;255;255m[1m  one-app-hmr ::[22m[39m %s",
    "[38;2;152;251;152mlocale[39m - [38;2;255;165;0m\\"en-us\\"[39m for module [38;2;255;165;0m\\"sample-module\\"[39m has been added",
  ],
]
`;

exports[`locale build watchLanguagePackFileEvents on change hot reload language pack  1`] = `
Array [
  Array [
    "[38;2;0;255;255m[1m  one-app-hmr ::[22m[39m %s",
    "[38;2;152;251;152mlocale[39m - [38;2;255;165;0m\\"en-us\\"[39m for module [38;2;255;165;0m\\"sample-module\\"[39m has been changed",
  ],
]
`;

exports[`locale build watchLanguagePackFileEvents on removed hot reload language pack 1`] = `
Array [
  Array [
    "[38;2;0;255;255m[1m  one-app-hmr ::[22m[39m %s",
    "[38;2;152;251;152mlocale[39m - [38;2;255;165;0m\\"en-us\\"[39m for module [38;2;255;165;0m\\"sample-module\\"[39m has been removed",
  ],
]
`;
